x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/DenseVector.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.vectors.query;
+
+import java.util.List;
+
+/**
+ * DenseVector value type for the painless.
+ */
+/* dotProduct, l1Norm, l2Norm, cosineSimilarity have three flavors depending on the type of the queryVector
+ * 1) float[], this is for the ScoreScriptUtils class bindings which have converted a List based query vector into an array
+ * 2) QueryVector, a wrapped List.  A painless script will typically use Lists since they are easy to pass as params and have an easy
+ *      literal syntax.  Working with Lists via QueryVector, instead of converting to a float[], trades off runtime operations against
+ *      memory pressure.  Dense Vectors may have high dimensionality, up to 2048.  Allocating a float[] per doc per script API call is
+ *      prohibitively expensive.
+ * 3) Object, the whitelisted method for the painless API.  Calls into the float[] or QueryVector version based on the
+        class of the argument and checks dimensionality. Wraps Lists as QueryVectors.
+ */
+public interface DenseVector {
+    float[] getVector();
+
+    float getMagnitude();
+
+    double dotProduct(float[] queryVector);
+
+    double dotProduct(QueryVector queryVector);
+
+    default double dotProduct(Object queryVector) {
+        if (queryVector instanceof float[] array) {
+            checkDimensions(getDims(), array.length);
+            return dotProduct(array);
+
+        } else if (queryVector instanceof QueryVector qv) {
+            checkDimensions(getDims(), qv.size());
+            return dotProduct(qv);
+
+        } else if (queryVector instanceof List<?> list) {
```

Comment:
Could this just be `instanceof List<Number>`? When we wrote the first version of vectors in painless, I think we could assume the param would have this type.

Reply:
We can't due to type erasure.  We can check `instanceof List<?>` and then do an unchecked cast below, adding `@SuppressWarnings("unchecked")`.

If they send in a bad type, they'll get a different reason message.  For the test
```yaml
  - do:
      catch: bad_request
      search:
        body:
          query:
            script_score:
              query: { "exists": { "field": "bdv" } }
              script:
                source: |
                  (int) field(params.field).get().l2Norm(params.query)
                params:
                  query: [1, 2, "three"]
                  field: bdv
```

the error will look like
```
        "failed_shards" : [
          {
            "shard" : 0,
            "index" : "test-index",
            "node" : "ueT4QigUSbK_ngYxjFfLZA",
            "reason" : {
              "type" : "script_exception",
              "reason" : "runtime error",
              "script_stack" : [
                "org.elasticsearch.xpack.vectors.query.BinaryDenseVector.l2Norm(BinaryDenseVector.java:127)",
                "org.elasticsearch.xpack.vectors.query.DenseVector.l2Norm(DenseVector.java:89)",
                "(int) field(params.field).get().l2Norm(params.query)
",
                "                                             ^---- HERE"
              ],
              "script" : "(int) field(params.field).get().l2Norm(params.query) ...",
              "lang" : "painless",
              "position" : {
                "offset" : 45,
                "start" : 0,
                "end" : 53
              },
              "caused_by" : {
                "type" : "class_cast_exception",
                "reason" : "class java.lang.String cannot be cast to class java.lang.Number (java.lang.String and java.lang.Number are in module java.base of loader 'bootstrap')",
                "stack_trace" : "java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Number (java.lang.String and java.lang.Number are in module java.base of loader 'bootstrap')
	at org.elasticsearch.xpack.vectors.query.BinaryDenseVector.l2Norm(BinaryDenseVector.java:127)

```

currently, the error looks like
```
        "failed_shards" : [
          {
            "shard" : 0,
            "index" : "test-index",
            "node" : "Fty-bHa-RC6HTekobvpLrw",
            "reason" : {
              "type" : "script_exception",
              "reason" : "runtime error",
              "script_stack" : [
                "org.elasticsearch.xpack.vectors.query.QueryVector.get(QueryVector.java:44)",
                "org.elasticsearch.xpack.vectors.query.BinaryDenseVector.l2Norm(BinaryDenseVector.java:126)",
                "org.elasticsearch.xpack.vectors.query.DenseVector.l2Norm(DenseVector.java:89)",
                "(int) field(params.field).get().l2Norm(params.query)
",
                "                                             ^---- HERE"
              ],
              "script" : "(int) field(params.field).get().l2Norm(params.query) ...",
              "lang" : "painless",
              "position" : {
                "offset" : 45,
                "start" : 0,
                "end" : 53
              },
              "caused_by" : {
                "type" : "illegal_argument_exception",
                "reason" : "Cannot treat [three] at index [2] of type [java.lang.String] as Number",
                "stack_trace" : "java.lang.IllegalArgumentException: Cannot treat [three] at index [2] of type [java.lang.String] as Number
	at org.elasticsearch.xpack.vectors.query.QueryVector.get(QueryVector.java:44)
```

ie the reason will be
> class java.lang.String cannot be cast to class java.lang.Number (java.lang.String and java.lang.Number are in module java.base of loader 'bootstrap')

vs

> Cannot treat [three] at index [2] of type [java.lang.String] as Number

Reply:
Reading quickly, I actually find the first one easier to understand, so happy to go with that if it helps simplify the code.

Reply:
Works for me.